Debug improvement

UnaKarlsen opened the pull request:
I have 2 small improvements to the debug enable functions.

Firstly, the existing grep search strings will return true even if the debug lines are commented out. Adding a carat symbol at the beginning to the grep search string causes it to check that the search string is at the beginning of the line, thus avoiding false positives.

The second improvement is to include any existing $OPTIONS in the docker config file. The default config file on Amazon Linux sets some ulimits, and folks may enable other options such as exposing the Docker API via TCP. by setting OPTIONS to "-D $OPTIONS" debug is enabled while retaining any customisation to the options, therefore not changing the environment in ways that may affect debugging. This still works even if debug has been enabled in the existing options.

UnaKarlsen committed:
added ^ to the grep match for ECS debug. This way the grep statement will not return true if the search string is found but commented out.

UnaKarlsen committed:
changed Docker debug option to "-D $OPTIONS" so that existing options are kept, while adding the debug flag. Likewise the Grep line has been updated to reflect this and updated to use ^ so that a commented out version will not cause a false positive match.

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -460,13 +460,13 @@ enable_docker_debug()
     amazon)
 
-      if [ -e /etc/sysconfig/docker ] && grep -q "OPTIONS=\"-D" /etc/sysconfig/docker
+      if [ -e /etc/sysconfig/docker ] && grep -q "^OPTIONS=\"-D \$OPTIONS" /etc/sysconfig/docker
```
I don't think we want the `\$OPTIONS` bit here. If the user hasn't done a similar appending as line 469, it will fail. Separately, I don't think we want to restrict ourselves to the line having to start with `OPTIONS`. Maybe `grep -q "^\s*OPTIONS=...` So we can match even if there is whitespace at the beginning of the line.

UnaKarlsen replied:
My intention was to check whether this script had enabled debug previously, or not. Even without the $OPTIONS bit, this grep check will only find instances where -D is at the beginning of the OPTIONS value. "^\s*OPTIONS=\".*\s*-D" would catch the debug switch even if placed after other options but risks a false positive if -D appears in a string for another option. I have done some testing and found that multiple instances of -D in the OPTIONS do not seem to have any adverse effects, so sticking with "^\s*OPTIONS=\"-D" seems to make the most sense.

Good point. Checking for white space is worth doing.

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -489,7 +489,7 @@ enable_ecs_agent_debug()
     amazon)
 
-      if [ -e /etc/ecs/ecs.config ] &&  grep -q "ECS_LOGLEVEL=debug" /etc/ecs/ecs.config
+      if [ -e /etc/ecs/ecs.config ] &&  grep -q "^ECS_LOGLEVEL=debug" /etc/ecs/ecs.config
```
Similar to above, lets match on `^\s*ECS_LOGLEVEL=debug` to catch people with indenting.

UnaKarlsen committed:
Updated Grep strings

Updated grep search strings for enabling debug mode, so that whitespace will be allowed before the search string.

SmilyBorg commented:
Updates to the Grep search strings have been made, to allow for whitespace prior to the search string.
$OPTIONS was also removed from the docker options search string as some users may not be appending strings.

SmilyBorg commented:
I have tested the updated script against amzn-ami-2017.03.f-amazon-ecs-optimized in both modes with both debug enabled and disabled prior to execution. All results were as expected.
